pkg/querier/querier.go

```diff
@@ -191,28 +202,86 @@ func (q querier) Select(sp *storage.SelectParams, matchers ...*labels.Matcher) (
 		}
 	}
 
-	return storage.NewMergeSeriesSet(result, nil), nil, nil
+	// we have all the sets from different sources (chunk from store, chunks from ingesters,
+	// time series from store and time series from ingesters).
+	return q.mergeSeriesSets(result), nil, nil
 }
 
 // LabelsValue implements storage.Querier.
 func (q querier) LabelValues(name string) ([]string, storage.Warnings, error) {
-	lv, err := q.distributor.LabelValuesForLabelName(q.ctx, model.LabelName(name))
-	return lv, nil, err
+	return q.primary.LabelValues(name)
 }
 
 func (q querier) LabelNames() ([]string, storage.Warnings, error) {
-	ln, err := q.distributor.LabelNames(q.ctx)
-	return ln, nil, err
+	return q.primary.LabelNames()
 }
 
-func (q querier) metadataQuery(matchers ...*labels.Matcher) (storage.SeriesSet, storage.Warnings, error) {
-	ms, err := q.distributor.MetricsForLabelMatchers(q.ctx, model.Time(q.mint), model.Time(q.maxt), matchers...)
-	if err != nil {
-		return nil, nil, err
+func (querier) Close() error {
+	return nil
+}
+
+func (q querier) mergeSeriesSets(sets []storage.SeriesSet) storage.SeriesSet {
+	// Here we deal with sets that are based on chunks and build single set from them.
+	// Remaining sets are merged with chunks-based one using storage.NewMergeSeriesSet
+
+	otherSets := []storage.SeriesSet(nil)
+	chunks := []chunk.Chunk(nil)
+
+	for _, set := range sets {
+		if !set.Next() {
+			// nothing in this set
+			continue
+		}
+
+		s := set.At()
+		if sc, ok := s.(SeriesWithChunks); ok {
+			chunks = append(chunks, sc.Chunks()...)
+
+			// iterate over remaining series in this set, and store chunks
+			// Here we assume that all remaining series in the set are also backed-up by chunks.
+			// If not, there will be panics.
+			for set.Next() {
+				s = set.At()
+				chunks = append(chunks, s.(SeriesWithChunks).Chunks()...)
+			}
+		} else {
+			otherSets = append(otherSets, &ignoreFirstNextSeriesSet{set: set})
+		}
+	}
+
+	if len(chunks) == 0 {
+		return storage.NewMergeSeriesSet(otherSets, nil)
+	}
+
+	// partitionChunks returns set with sorted series, so it can be used by NewMergeSeriesSet
+	chunksSet := partitionChunks(chunks, q.mint, q.maxt, q.chunkIterFn)
+
+	if len(otherSets) == 0 {
+		return chunksSet
 	}
-	return metricsToSeriesSet(ms), nil, nil
+
+	otherSets = append(otherSets, chunksSet)
+	return storage.NewMergeSeriesSet(otherSets, nil)
 }
 
-func (querier) Close() error {
-	return nil
+// This series set ignores first 'Next' call and simply returns true.
+type ignoreFirstNextSeriesSet struct {
+	firstNextCalled bool
+	set             storage.SeriesSet
+}
+
+func (pss *ignoreFirstNextSeriesSet) Next() bool {
+	if pss.firstNextCalled {
+		return pss.set.Next()
+	}
+	pss.firstNextCalled = true
+	return true
```

Comment:
What if the series set has only 1 item, so `Next()` is expected to return `false` on the first call? If this is a real issue, please cover this case with a unit test.

Reply:
> What if the series set has only 1 item, so `Next()` is expected to return `false` on the first call? If this is a real issue, please cover this case with a unit test.

I don't understand what you mean. Reason for this is that we call `set.Next()` once for every set to check if it is chunk-based or not. It if isn't, then we want to make sure that given set returns same result from `At()` call again to the query engine. Hence wrapping into this struct.

Now looking at it, I think I should cache that first `At()` result and simply return it again, instead of relying on set returning the same value twice.

Reply:
> Now looking at it, I think I should cache that first At() result and simply return it again, instead of relying on set returning the same value twice.

What's wrong with calling `At()` twice for the same entry? If there's something wrong (or potentially wrong) would be better to add a comment to explain it. I'm missing it.

Reply:
Since we already have the value returned by first `At()`, there is no need to call it again. There are implementations of this function that do extra work each time `At()` is called (eg. `mergeSeriesSet` does the merging again, `dedupSeriesSet` does the deduplication again, etc.), so it seemed like a good idea.

Reply:
I've changed the comment on `seriesSetWithFirstSeries` to mention this.

Reply:
> do extra work each time At() is called

Ok. I was wondering if it was an optimization or bug fix. You confirmed it's just an optimization.